src/irmin/perms.ml

```diff
@@ -0,0 +1,49 @@
+(** Types representing {i permissions} ['perms] for performing operations on a
+    certain type ['perms t].
+
+    They are intended to be used as phantom parameters of the types that they
+    control access to. As an example, consider the following type of references
+    with permissions:
+
+    {[
+      module Ref : sig
+        type (+'a, -'perms) t
+
+        val create : 'a -> ('a, read_write) t
+        val get : ('a, [> read ]) t -> 'a
+        val set : ('a, [> write ]) t -> 'a -> unit
+      end
+    ]}
+
+    This type allows references to be created with arbitrary read-write access.
+    One can then create weaker views onto the reference – with access to fewer
+    operations – by upcasting:
```

Comment:
I think all type coercions in OCaml are upcasting, since there is no runtime type information with which to downcast. i.e. `t :> u` requires that `t` is a subtype of `u`.

Perhaps it's best to avoid this altogether and use `coercing` instead.

Reply:
👍 I had my up/downcast definitions upside down